### PR TITLE
noripple_check implementation of new RPC system 

### PR DESCRIPTION
**Pull request by cindyyan317**
noripple_check implementation of new RPC system

### CODE REVIEW

File: src/rpc/ngHandlers/NoRippleCheck.cpp [hunk 1]
```diff
@@ -0,0 +1,221 @@
...
+        input.ledgerHash,
+        input.ledgerIndex,
+        range->maxSequence);
+
+    if (auto status = std::get_if<RPC::Status>(&lgrInfoOrStatus))
+        return Error{*status};
+
+    auto const lgrInfo = std::get<ripple::LedgerInfo>(lgrInfoOrStatus);
+    auto const accountID = RPC::accountFromStringStrict(input.account);
+    auto keylet = ripple::keylet::account(*accountID).key;
```
Comment (godexsoft) [inline src/rpc/ngHandlers/NoRippleCheck.cpp]: Could this also be `const`?

File: src/rpc/ngHandlers/NoRippleCheck.cpp [hunk 2]
```diff
@@ -0,0 +1,221 @@
...
+        sharedPtrBackend_->fetchLedgerObject(keylet, lgrInfo.seq, yield);
+    if (!accountObj)
+        return Error{RPC::Status{
+            RPC::RippledError::rpcACT_NOT_FOUND, "accountNotFound"}};
+
+    ripple::SerialIter it{accountObj->data(), accountObj->size()};
+    ripple::SLE sle{it, keylet};
+    std::uint32_t accountSeq = sle.getFieldU32(ripple::sfSequence);
+
+    bool bDefaultRipple =
```
Comment (godexsoft) [inline src/rpc/ngHandlers/NoRippleCheck.cpp]: Looks like can be `const` also.

File: src/rpc/ngHandlers/NoRippleCheck.cpp [hunk 3]
```diff
@@ -0,0 +1,221 @@
...
+NoRippleCheckHandler::Input
+tag_invoke(
+    boost::json::value_to_tag<NoRippleCheckHandler::Input>,
+    boost::json::value const& jv)
+{
+    auto const& jsonObject = jv.as_object();
+    NoRippleCheckHandler::Input input;
+    input.account = jv.at("account").as_string().c_str();
+    input.roleGateway = jv.at("role").as_string() == "gateway";
+    if (jsonObject.contains("limit"))
```
Comment (godexsoft) [inline src/rpc/ngHandlers/NoRippleCheck.cpp]: minor: Consider using JStrings for all these also

File: src/rpc/ngHandlers/NoRippleCheck.h [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+    NoRippleCheckHandler(
+        std::shared_ptr<BackendInterface> const& sharedPtrBackend)
+        : sharedPtrBackend_(sharedPtrBackend)
+    {
+    }
+
+    RpcSpecConstRef
+    spec() const
+    {
+        static const RpcSpec rpcSpec = {
```
Comment (godexsoft) [inline src/rpc/ngHandlers/NoRippleCheck.h]: minor: This was my original mistake but i tried to fix in new PRs so consider also fixing: `static auto const rpcSpec = RpcSpec{...}`

File: src/rpc/ngHandlers/NoRippleCheck.h [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+            {"account", validation::Required{}, validation::AccountValidator},
```
Comment (godexsoft) [inline src/rpc/ngHandlers/NoRippleCheck.h]: minor: Can probably utilise JStrings here too

**Review by godexsoft — CHANGES_REQUESTED** [review]
Leaving a few suggestions. Overall very good 👍

File: src/rpc/ngHandlers/NoRippleCheck.cpp [hunk 6]
```diff
@@ -0,0 +1,220 @@
...
+    auto const getBaseTx = [&](ripple::AccountID const& accountID,
+                               std::uint32_t accountSeq) {
+        boost::json::object tx;
+        tx[JS(Sequence)] = accountSeq;
+        tx[JS(Account)] = ripple::toBase58(accountID);
+        tx[JS(Fee)] = RPC::toBoostJson(fees->units.jsonClipped());
+        return tx;
+    };
+
+    if (bDefaultRipple & !input.roleGateway)
```
Comment (godexsoft) [inline src/rpc/ngHandlers/NoRippleCheck.cpp]: Although this yields the same result in this case, I suppose the intention was to use `&&`, not `&` here.

File: src/rpc/ngHandlers/NoRippleCheck.cpp [hunk 7]
```diff
@@ -0,0 +1,220 @@
...
+    {
+        output.problems.push_back(
+            "You appear to have set your default ripple flag even though "
+            "you "
+            "are not a gateway. This is not recommended unless you are "
+            "experimenting");
+    }
+    else if (input.roleGateway & !bDefaultRipple)
```
Comment (godexsoft) [inline src/rpc/ngHandlers/NoRippleCheck.cpp]: Same here.

File: src/rpc/ngHandlers/NoRippleCheck.cpp [hunk 8]
```diff
@@ -0,0 +1,220 @@
...
+                    (bLow ? ripple::lsfLowNoRipple : ripple::lsfHighNoRipple);
+
+                std::string problem;
+                bool needFix = false;
+                if (bNoRipple & input.roleGateway)
+                {
+                    problem = "You should clear the no ripple flag on your ";
+                    needFix = true;
+                }
+                else if (!bNoRipple & !input.roleGateway)
```
Comment (godexsoft) [inline src/rpc/ngHandlers/NoRippleCheck.cpp]: And here.

File: src/rpc/ngHandlers/NoRippleCheck.cpp [hunk 9]
```diff
@@ -0,0 +1,220 @@
...
+            {
+                bool const bLow = accountID ==
+                    ownedItem.getFieldAmount(ripple::sfLowLimit).getIssuer();
+
+                bool const bNoRipple = ownedItem.getFieldU32(ripple::sfFlags) &
+                    (bLow ? ripple::lsfLowNoRipple : ripple::lsfHighNoRipple);
+
+                std::string problem;
+                bool needFix = false;
+                if (bNoRipple & input.roleGateway)
```
Comment (godexsoft) [inline src/rpc/ngHandlers/NoRippleCheck.cpp]: And here.

File: src/rpc/ngHandlers/NoRippleCheck.cpp [hunk 10]
```diff
@@ -0,0 +1,220 @@
...
+}
+
+NoRippleCheckHandler::Input
+tag_invoke(
+    boost::json::value_to_tag<NoRippleCheckHandler::Input>,
+    boost::json::value const& jv)
+{
+    auto const& jsonObject = jv.as_object();
+    NoRippleCheckHandler::Input input;
+    input.account = jv.at(JS(account)).as_string().c_str();
```
Comment (godexsoft) [inline src/rpc/ngHandlers/NoRippleCheck.cpp]: I wonder, why not use jsonObject here too?

File: src/rpc/ngHandlers/NoRippleCheck.cpp [hunk 11]
```diff
@@ -0,0 +1,220 @@
...
+    input.roleGateway = jv.at(JS(role)).as_string() == "gateway";
+    if (jsonObject.contains(JS(limit)))
+    {
+        input.limit = jv.at(JS(limit)).as_int64();
```
Comment (godexsoft) [inline src/rpc/ngHandlers/NoRippleCheck.cpp]: And also here and in similar situations below.

File: src/rpc/ngHandlers/NoRippleCheck.h [hunk 12]
```diff
@@ -0,0 +1,99 @@
...
+{
+    std::shared_ptr<BackendInterface> sharedPtrBackend_;
+
+public:
+    struct Output
+    {
+        std::string ledgerHash;
+        uint32_t ledgerIndex;
+        std::vector<std::string> problems;
+        std::optional<boost::json::array> transactions;
```
Comment (godexsoft) [inline src/rpc/ngHandlers/NoRippleCheck.h:43]: Please add `TODO` to get rid of json in here

File: unittests/rpc/handlers/NoRippleCheckTest.cpp [hunk 13]
```diff
@@ -0,0 +1,737 @@
...
+    NormalPathRoleUserDefaultRippleUnsetTrustLineNoRippleUnSet)
+{
+    static auto constexpr seq = 30;
+    static auto constexpr expectedOutput =
+        R"({
+            "ledger_hash":"4BC50C9B0D8515D3EAAE1E74B29A95804346C491EE1A95BF25E4AAB854A6A652",
+            "ledger_index":30,
+            "problems":[
+                "You should probably set the no ripple flag on your USD line to rLEsXccBGNR3UPuPu2hUXPjziKC3qKSBun",
+                "You should probably set the no ripple flag on your USD line to rLEsXccBGNR3UPuPu2hUXPjziKC3qKSBun"
```
Comment (godexsoft) [inline unittests/rpc/handlers/NoRippleCheckTest.cpp:325]: Is it really supposed to output the same problem twice or is this a bug in the original code?

**Review by godexsoft — CHANGES_REQUESTED** [review]
Found some more little things to improve and/or check 👍

File: unittests/rpc/handlers/NoRippleCheckTest.cpp (hunk 13 above)
Comment (cindyyan317) [reply unittests/rpc/handlers/NoRippleCheckTest.cpp:325]: Mock the same trustline twice , so it should be two problems here

File: src/rpc/ngHandlers/NoRippleCheck.cpp (hunk 10 above)
Comment (cindyyan317) [reply src/rpc/ngHandlers/NoRippleCheck.cpp]: json object and value both have at() function. jsonObject here is mainly for contains() function.

**Review by godexsoft — DISMISSED** [review]
LGTM! 🚀

**Review by godexsoft — APPROVED** [review]
Go Go 🔥

**Review by godexsoft — APPROVED** [review]
Go go

**Pull request merged**